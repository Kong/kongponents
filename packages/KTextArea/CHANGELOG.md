# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 6.1.4 (2021-11-17)


### Bug Fixes

* general fixes ([#415](https://github.com/Kong/kongponents/issues/415)) ([5b4bb48](https://github.com/Kong/kongponents/commit/5b4bb484d74e75779d39cf901b2cdaa23818e8e7))





## 6.1.3 (2021-11-17)

**Note:** Version bump only for package @kongponents/ktextarea
