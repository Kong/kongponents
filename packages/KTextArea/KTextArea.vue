<template>
  <div class="k-input-wrapper mb-2">
    <textarea
      v-if="!label"
      v-bind="$attrs"
      :value="currValue ? currValue : value"
      :rows="rows"
      :cols="cols"
      class="form-control k-input style-body-lg"
      @input="e => {
        $emit('textarea', e.target.value),
        currValue = e.target.value

        if (!disableCharacterLimit && currValue.length > characterLimit) {
          $emit('char-limit-exceed', `Character limit of ${characterLimit} execeeded by ${currValue.length - characterLimit} characters`)
        }
      }"
      v-on="listeners"
    />

    <div
      v-else
      class="col-md-4 mt-5">
      <div class="text-on-input">
        <label
          :for="textAreaId"
          :class="{ focused: isFocused, hovered: isHovered }">
          <span>{{ label }}</span>
        </label>
        <textarea
          v-bind="$attrs"
          :id="textAreaId"
          :value="currValue ? currValue : value"
          :rows="rows"
          :cols="cols"
          class="form-control k-input style-body-lg"
          @input="e => {
            $emit('textarea', e.target.value),
            currValue = e.target.value

            if (!disableCharacterLimit && currValue.length > characterLimit) {
              $emit('char-limit-exceed', `Character limit of ${characterLimit} execeeded by ${currValue.length - characterLimit} characters`)
            }
          }"
          @mouseenter="() => isHovered = true"
          @mouseleave="() => isHovered = false"
          @focus="() => isFocused = true"
          @blur="() => isFocused = false"
          v-on="listeners" />
      </div>
    </div>

    <div
      v-if="!disableCharacterLimit"
      :class="{ 'over-char-limit': currValue.length > characterLimit }"
      class="char-limit type-sm color-black-45 mt-2">
      {{ currValue.length || value.length }} / {{ characterLimit }}
    </div>
  </div>
</template>

<script>
import { uuid } from 'vue-uuid'
const CHARACTER_LIMIT = 2048

export default {
  name: 'KTextArea',
  inheritAttrs: false,
  props: {
    value: {
      type: String,
      default: ''
    },
    label: {
      type: String,
      default: ''
    },
    characterLimit: {
      type: Number,
      default: CHARACTER_LIMIT
    },
    disableCharacterLimit: {
      type: Boolean,
      default: false
    },
    rows: {
      type: Number,
      default: 5
    },
    cols: {
      type: Number,
      default: 52
    },
    /**
     * Test mode - for testing only, strips out generated ids
     */
    testMode: {
      type: Boolean,
      default: false
    }
  },
  data: function () {
    return {
      currValue: '', // We need this so that we don't lose the updated value on hover/blur event with label
      isFocused: false,
      isHovered: false
    }
  },
  computed: {
    textAreaId () {
      return this.$attrs.id ? this.$attrs.id : !this.testMode ? uuid.v1() : 'test-textArea-id-1234'
    },
    listeners () {
      const listeners = { ...this.$listeners }

      // use @input in template for v-model support
      delete listeners['input']

      return listeners
    }
  }
}
</script>

<style lang="scss" scoped>
@import '~@kongponents/styles/_variables.scss';
@import '~@kongponents/styles/forms/_inputs.scss';

.k-input-wrapper {
  width: fit-content;
  display: grid;

  textarea.form-control {
    font-family: var(--font-family-sans);
    resize: none;

    &::placeholder {
      color: var(--grey-500);
    }

    &:hover {
      color: var(--grey-600);
    }

    &:hover::placeholder {
      color: var(--grey-600);
    }

    &:focus::placeholder {
      color: transparent;
    }
  }

  .char-limit {
    margin-left: auto;
  }

  .over-char-limit {
    color: var(--red-600);
  }
}
</style>
