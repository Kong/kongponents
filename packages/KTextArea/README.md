# @kongponents/ktextarea

[![](https://img.shields.io/npm/v/@kongponents/ktextarea.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/ktextarea)

```vue
<KTextArea :description="'hello world'">
  Hello from a slot
</KTextArea>
```
