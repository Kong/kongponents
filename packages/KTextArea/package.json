{
  "name": "@kongponents/ktextarea",
  "version": "6.32.16",
  "description": "Text area component",
  "main": "dist/KTextArea.umd.min.js",
  "componentName": "KTextArea",
  "source": "KTextArea.vue",
  "files": [
    "dist",
    "KTextArea.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/klabel": "^6.32.16",
    "@kongponents/styles": "^6.32.16",
    "vue-uuid": "^2.0.2"
  }
}
