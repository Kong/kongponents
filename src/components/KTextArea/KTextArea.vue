<template>
  <div class="k-input-wrapper mb-2">
    <textarea
      v-if="!label"
      v-bind="$attrs"
      :value="currValue ? currValue : modelValue"
      :rows="rows"
      :cols="cols"
      class="form-control k-input style-body-lg"
      @input="inputHandler"
    />

    <div
      v-else-if="label && overlayLabel"
      class="mt-5"
    >
      <div class="text-on-input">
        <label
          :for="textAreaId"
          v-bind="labelAttributes"
          :class="{ focused: isFocused, hovered: isHovered }"
        >
          <span>{{ label }}</span>
        </label>
        <textarea
          v-bind="$attrs"
          :id="textAreaId"
          :value="currValue ? currValue : modelValue"
          :rows="rows"
          :cols="cols"
          class="form-control k-input style-body-lg"
          @input="inputHandler"
          @mouseenter="() => isHovered = true"
          @mouseleave="() => isHovered = false"
          @focus="() => isFocused = true"
          @blur="() => isFocused = false"
        />
      </div>
    </div>

    <div
      v-else
      class="mt-5"
    >
      <KLabel
        :for="textAreaId"
        v-bind="labelAttributes"
      >
        {{ label }}
      </KLabel>
      <textarea
        v-bind="$attrs"
        :id="textAreaId"
        :value="currValue ? currValue : modelValue"
        :rows="rows"
        :cols="cols"
        class="form-control k-input style-body-lg"
        @input="inputHandler"
        @mouseenter="() => isHovered = true"
        @mouseleave="() => isHovered = false"
        @focus="() => isFocused = true"
        @blur="() => isFocused = false"
      />
    </div>

    <div
      v-if="!disableCharacterLimit"
      :class="{ 'over-char-limit': currValue.length > characterLimit }"
      class="char-limit type-sm color-black-45 mt-2"
    >
      {{ currValue.length || modelValue.length }} / {{ characterLimit }}
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, computed, watch } from 'vue'
import KLabel from '@/components/KLabel/KLabel.vue'
import { v1 as uuidv1 } from 'uuid'

const CHARACTER_LIMIT = 2048

export default defineComponent({
  name: 'KTextArea',
  components: { KLabel },
  inheritAttrs: false,
  props: {
    modelValue: {
      type: String,
      default: '',
    },
    label: {
      type: String,
      default: '',
    },
    overlayLabel: {
      type: Boolean,
      default: false,
    },
    labelAttributes: {
      type: Object,
      default: () => ({}),
    },
    characterLimit: {
      type: Number,
      default: CHARACTER_LIMIT,
    },
    disableCharacterLimit: {
      type: Boolean,
      default: false,
    },
    rows: {
      type: Number,
      default: 5,
    },
    cols: {
      type: Number,
      default: 52,
    },
    /**
     * Test mode - for testing only, strips out generated ids
     */
    testMode: {
      type: Boolean,
      default: false,
    },
  },
  emits: ['input', 'update:modelValue', 'char-limit-exceeded'],
  setup(props, { attrs, emit }) {
    const currValue = ref('') // We need this so that we don't lose the updated value on hover/blur event with label
    const isFocused = ref(false)
    const isHovered = ref(false)

    const textAreaId = computed((): string => (attrs.id ? String(attrs.id) : !props.testMode ? uuidv1() : 'test-textArea-id-1234'))

    const charLimitExceeded = computed((): boolean => !props.disableCharacterLimit && currValue.value.length > props.characterLimit)

    const inputHandler = (e: any) => {
      // this 'input' and 'update:modelValue' events must be emitted for v-model binding to work properly
      emit('input', e.target.value)
      emit('update:modelValue', e.target.value)
      currValue.value = e.target.value
    }

    watch(charLimitExceeded, (newVal, oldVal) => {
      if (newVal !== oldVal) {
        emit('char-limit-exceeded', {
          value: currValue.value,
          length: currValue.value.length,
          characterLimit: props.characterLimit,
          limitExceeded: newVal,
        })
      }
    })

    return {
      currValue,
      isFocused,
      isHovered,
      textAreaId,
      charLimitExceeded,
      inputHandler,
    }
  },
})
</script>

<style lang="scss" scoped>
@import '@/styles/variables';

.k-input-wrapper {
  width: fit-content;
  display: grid;

  textarea.k-input {
    -webkit-appearance: none;
  }

  textarea.form-control {
    font-family: var(--font-family-sans);
    resize: none;

    &::placeholder {
      color: var(--KInputPlaceholderColor, var(--grey-500, color(grey-500)));
    }

    &:hover {
      color: var(--grey-600);
    }

    &:hover::placeholder {
      color: var(--KInputPlaceholderColor, var(--grey-600, color(grey-600)));
    }

    &:focus::placeholder {
      color: transparent;
    }
  }

  .char-limit {
    margin-left: auto;
  }

  .over-char-limit {
    color: var(--red-600);
  }
}
</style>
