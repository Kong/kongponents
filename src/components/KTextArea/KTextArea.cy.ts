import { mount } from 'cypress/vue'
import KTextArea from '@/components/KTextArea/KTextArea.vue'

/**
 * ALL TESTS MUST USE testMode: true
 * We generate unique IDs for reference by aria properties. Debug mode strips these out
 * allowing for successful snapshot verification.
 * props: {
 *   testMode: true
 * }
 */

describe('KTextArea', () => {
  it('renders text when value is passed', () => {
    const value = 'Howdy!'
    mount(KTextArea, {
      props: {
        testMode: true,
        modelValue: value, // v-model
      },
    })

    cy.get('textarea').should('have.value', value)
  })

  it('renders label when value is passed', () => {
    const labelText = 'A Label!'
    mount(KTextArea, {
      props: {
        testMode: true,
        label: labelText,
      },
    })

    cy.get('.k-label').should('contain.text', labelText)
  })

  it('renders label with labelAttributes applied', () => {
    const labelText = 'A Label'
    mount(KTextArea, {
      props: {
        testMode: true,
        label: labelText,
        labelAttributes: {
          info: 'some help text',
        },
      },
    })

    cy.get('.k-label').should('contain.text', labelText)
    cy.get('.k-label .tooltip-trigger-icon').should('be.visible')
  })

  it('renders overlayed label when value is passed', () => {
    const labelText = 'A Label'
    mount(KTextArea, {
      props: {
        testMode: true,
        label: labelText,
        overlayLabel: true,
      },
    })

    cy.get('.text-on-input label').should('contain.text', labelText)
  })

  it('renders an asterisk when `overlayLabel` is true and `required` attr is set', () => {
    const label = 'A label'
    mount(KTextArea, {
      props: {
        testMode: true,
        label,
        overlayLabel: true,
      },
      attrs: {
        required: true,
      },
    })

    cy.get('.text-on-input  .is-required').should('exist')
  })

  it('renders textarea when rows and cols are passed in', () => {
    mount(KTextArea, {
      props: {
        testMode: true,
        rows: 2,
        cols: 15,
      },
    })

    cy.get('textarea').should('be.visible')
  })

  it('reacts to text changes', () => {
    const value1 = 'hey'
    const value2 = 'hey, dude'
    mount(KTextArea, {
      props: {
        testMode: true,
        modelValue: value1, // v-model
      },
    })

    cy.get('textarea').should('have.value', value1)
    cy.get('textarea').clear()
    cy.get('textarea').type(value2).then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'input')
      const emitArray = Cypress.vueWrapper.emitted('input')
      cy.wrap(String(emitArray?.[emitArray.length - 1])).should('eq', value2)
    })
    cy.get('textarea').should('have.value', value2)
  })

  it('can configure character limit', () => {
    const charLimit = 500
    mount(KTextArea, {
      props: {
        testMode: true,
        characterLimit: charLimit,
      },
    })

    cy.get('.char-limit').should('contain.text', charLimit)
  })

  it('should have style when value exceeds the character limit', () => {
    const charLimit = 20
    mount(KTextArea, {
      props: {
        testMode: true,
        characterLimit: charLimit,
      },
    })

    cy.get('textarea').type('a'.repeat(charLimit + 1))
    cy.get('textarea').should('have.value', 'a'.repeat(charLimit + 1))
    cy.get('div.over-char-limit').should('be.visible')
  })

  it('should allow disabling character limit', () => {
    mount(KTextArea, {
      props: {
        testMode: true,
        disableCharacterLimit: true,
      },
    })

    cy.get('.char-limit').should('not.exist')
  })

  it('should have `is-resizable` class when is-resizable prop is enabled', () => {
    mount(KTextArea, {
      props: {
        isResizable: true,
      },
    })

    cy.get('textarea').should('have.class', 'is-resizable')
  })
})
