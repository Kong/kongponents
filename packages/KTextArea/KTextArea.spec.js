import { mount } from '@vue/test-utils'
import KTextArea from '@/KTextArea/KTextArea'

/**
 * ALL TESTS MUST USE testMode: true
 * We generate unique IDs for reference by aria properties. Debug mode strips these out
 * allowing for successful snapshot verification.
 * propsData: {
 *   testMode: true
 * }
 */

describe('KTextArea', () => {
  it('renders text when value is passed', () => {
    const wrapper = mount(KTextArea, {
      propsData: {
        testMode: true,
        value: 'Hello' // e.g. v-model
      }
    })

    expect(wrapper.find('textarea').element.value).toBe('Hello')
  })

  it('renders label when value is passed', () => {
    const wrapper = mount(KTextArea, {
      propsData: {
        testMode: true,
        label: 'A Label!'
      }
    })

    expect(wrapper.find('.text-on-input label').element.innerHTML).toContain('A Label!')
  })

  it('renders textarea when rows and cols are passed in', () => {
    const wrapper = mount(KTextArea, {
      propsData: {
        testMode: true,
        rows: 2,
        cols: 15
      }
    })

    expect(wrapper.find('textarea').exists()).toBeTruthy()
  })

  it('reacts to text changes', () => {
    const wrapper = mount(KTextArea, {
      propsData: {
        testMode: true,
        value: 'hey'
      }
    })
    const textarea = wrapper.find('textarea')

    expect(textarea.element.value).toBe('hey')
    textarea.setValue('hey, dude')

    expect(wrapper.emitted().textarea).toHaveLength(1)
    expect(wrapper.emitted().textarea[0]).toEqual(['hey, dude'])
    expect(textarea.element.value).toBe('hey, dude')
  })

  it('can configure character limit', () => {
    const charLimit = 500
    const wrapper = mount(KTextArea, {
      propsData: {
        testMode: true,
        characterLimit: charLimit
      }
    })

    expect(wrapper.find('.char-limit').html()).toContain(charLimit)
  })

  it('should have style when value exceeds the character limit', () => {
    const value = 'abc'.repeat(683)
    const wrapper = mount(KTextArea, {
      propsData: {
        testMode: true,
        value
      }
    })
    const textarea = wrapper.find('textarea')

    textarea.setValue(value)
    expect(wrapper.find('textarea').element.value).toBe(value)
    expect(wrapper.find('div.over-char-limit').exists()).toBe(true)
  })

  it('should allow disabling character limit', () => {
    const wrapper = mount(KTextArea, {
      propsData: {
        testMode: true,
        disableCharacterLimit: true
      }
    })

    expect(wrapper.find('.char-limit').exists()).toBe(false)
  })

  it('matches snapshot', () => {
    const wrapper = mount(KTextArea, {
      propsData: {
        testMode: true,
        value: 'Full Name',
        placeholder: 'I am a placeholder',
        name: 'custom-textArea-name',
        id: 'custom-textArea-id',
        'data-testid': 'custom-textArea'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
