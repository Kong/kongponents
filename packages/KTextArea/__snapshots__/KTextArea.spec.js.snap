// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KTextArea matches snapshot 1`] = `
<div class="k-input-wrapper" placeholder="I am a placeholder" name="custom-textArea-name" id="custom-textArea-id" data-testid="custom-textArea"><textarea rows="5" cols="52" placeholder="I am a placeholder" name="custom-textArea-name" id="custom-textArea-id" data-testid="custom-textArea" class="form-control k-input style-body-lg"></textarea>
  <div class="type-sm color-black-45 float-right">
    9 / 2048
  </div>
</div>
`;
